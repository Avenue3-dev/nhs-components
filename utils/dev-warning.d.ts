/**
 * Print a warning when using experimental components when not in production mode.
 *
 * @param component name of the experimental component.
 */
export declare const useExperimentalWarning: (component: 'tabs' | 'character-count') => void;
/**
 * Print a warning message when not in production mode.
 *
 * @param message message to print when not it production.
 */
export declare const useDevWarning: (message: string) => void;
