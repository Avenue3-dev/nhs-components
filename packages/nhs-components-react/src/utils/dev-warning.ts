const prod = process.env.NODE_ENV === 'production'

export const experimentalWarning = (
  component: 'tabs' | 'character-count',
): void => {
  if (prod) {
    return
  }

  console.warn(`ATTENTION
  This component is tagged as experimental and therefore could change without notice. See more details about this component at https://service-manual.nhs.uk/design-system/components/${component}
  `)
}
