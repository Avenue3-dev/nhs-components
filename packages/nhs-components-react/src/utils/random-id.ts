export const generateRandomString = (length = 5): string => {
  const rand = Math.random() + 1

  return rand.toString(36).substring(2, length + 2)
}

export const generateRandomName = (prefix?: string): string => {
  const randString = generateRandomString()

  return prefix ? `${prefix}_${randString}` : randString
}

export const generateRandomId = generateRandomName
