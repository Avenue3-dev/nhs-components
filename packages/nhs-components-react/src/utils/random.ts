export const generateRandomString = (length: number = 8): string => {
  return [...Array<string>(length)]
    .map(() => Math.random().toString(36)[2])
    .join('')
}
