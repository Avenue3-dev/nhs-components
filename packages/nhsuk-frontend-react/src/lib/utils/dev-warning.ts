const prod = process.env.NODE_ENV === 'production'

/**
 * Print a warning when using experimental components when not in production mode.
 *
 * @param component name of the experimental component.
 */
export const useExperimentalWarning = (
  component: 'tabs' | 'character-count',
): void => {
  useDevWarning(`This component is tagged as experimental and therefore could change without notice.

  See more details about this component at https://service-manual.nhs.uk/design-system/components/${component}
  `)
}

/**
 * Print a warning message when not in production mode.
 *
 * @param message message to print when not it production.
 */
export const useDevWarning = (message: string): void => {
  if (prod) {
    return
  }

  console.warn(message)
}
