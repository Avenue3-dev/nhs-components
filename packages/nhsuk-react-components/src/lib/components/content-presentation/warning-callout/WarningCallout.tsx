import { HTMLProps } from 'react'
import clsx from 'clsx'
import { VisuallyHiddenProps } from '@/types/visually-hidden'
import { AsElementHeadingProps } from '@/types/heading'

type WarningCallout = {
  Label: typeof Label
} & React.FC<HTMLProps<HTMLDivElement>>

type WarningCalloutLabelProps = AsElementHeadingProps &
  HTMLProps<HTMLHeadingElement> &
  Partial<VisuallyHiddenProps>

/**
 * The label for the `WarningCallout` component.
 *
 * @example
 * ```tsx
 * <WarningCallout.Label>Important</WarningCallout.Label>
 * ```
 */
const Label: React.FC<WarningCalloutLabelProps> = ({
  children,
  className,
  asElement: Component = 'h3',
  visuallyHiddenText = 'Important',
  ...rest
}) => {
  return (
    <Component
      className={clsx('nhsuk-warning-callout__label', className)}
      {...rest}
    >
      <span className="nhsuk-u-visually-hidden">{visuallyHiddenText}</span>
      {children}
    </Component>
  )
}

/**
 * Use a warning callout to help users identify and understand warning content on the page, even if they do not read the whole page.
 *
 * For more information on when to use this component, go to the [NHS Service Manuals website](https://service-manual.nhs.uk/design-system/components/warning-callout).
 *
 * @link https://service-manual.nhs.uk/design-system/components/warning-callout
 *
 * @example
 * ```tsx
 * <WarningCallout>
 *  <WarningCallout.Label>Important</WarningCallout.Label>
 *  <p>
 *   If you have symptoms of coronavirus (COVID-19), you must self-isolate
 *  immediately.
 *  </p>
 * </WarningCallout>
 * ```
 */
const WarningCallout: WarningCallout = ({ children, className, ...rest }) => {
  return (
    <div className={clsx('nhsuk-warning-callout', className)} {...rest}>
      {children}
    </div>
  )
}

WarningCallout.displayName = 'WarningCallout'
Label.displayName = 'WarningCallout.Label'

WarningCallout.Label = Label

export default WarningCallout
