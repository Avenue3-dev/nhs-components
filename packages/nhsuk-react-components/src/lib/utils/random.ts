export const generateRandomString = (length = 8): string => {
  return [...Array<string>(length)]
    .map(() => Math.random().toString(36)[2])
    .join('')
}
