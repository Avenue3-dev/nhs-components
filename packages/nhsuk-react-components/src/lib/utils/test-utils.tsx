/* eslint-disable import/export */
import {
  RenderOptions,
  RenderResult,
  cleanup,
  render,
} from '@testing-library/react'
import { MemoryRouter as Router } from 'react-router-dom'
import { afterEach } from 'vitest'

afterEach(() => {
  cleanup()
})

const customRender = (
  ui: React.ReactElement,
  options: RenderOptions = {},
): RenderResult => {
  return render(ui, {
    // wrap provider(s) here if needed
    wrapper: ({ children }) => <Router>{children}</Router>,
    ...options,
  })
}

export * from '@testing-library/react'
export { default as userEvent } from '@testing-library/user-event'
// override render export
export { customRender as render }
