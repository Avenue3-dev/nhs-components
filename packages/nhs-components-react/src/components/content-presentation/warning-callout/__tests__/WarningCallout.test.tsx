import { render, screen } from '@/utils/test-utils'
import { describe, expect, test } from 'vitest'
import WarningCallout from '..'

describe('WarningCallout', () => {
  test('should match snapshot', () => {
    render(
      <WarningCallout data-testid="warning-callout">
        <WarningCallout.Label>Important information</WarningCallout.Label>
        <p>Lorem ipsum dolor, sit amet consectetur adipisicing elit. Temporibus quidem unde nobis, rem consequuntur cum, illo, tempore error ipsum dolor fugit fuga excepturi id! Molestias veritatis consequuntur iste deserunt dicta!</p>
      </WarningCallout>
    )

    expect(screen.getByTestId('warning-callout')).toMatchSnapshot()
  })
})
