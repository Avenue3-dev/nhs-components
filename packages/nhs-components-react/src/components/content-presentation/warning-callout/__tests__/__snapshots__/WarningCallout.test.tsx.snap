// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WarningCallout > should match snapshot 1`] = `
<div
  class="nhsuk-warning-callout"
  data-testid="warning-callout"
>
  <h3
    class="nhsuk-warning-callout__label"
  >
    <span
      class="nhsuk-u-visually-hidden"
    >
      Important
    </span>
    Important information
  </h3>
  <p>
    Lorem ipsum dolor, sit amet consectetur adipisicing elit. Temporibus quidem unde nobis, rem consequuntur cum, illo, tempore error ipsum dolor fugit fuga excepturi id! Molestias veritatis consequuntur iste deserunt dicta!
  </p>
</div>
`;
